src/lib/optimizer/strategy/join_predicate_ordering_rule.cpp

```diff
@@ -0,0 +1,69 @@
+#include "join_predicate_ordering_rule.hpp"
+
+#include "cost_estimation/abstract_cost_estimator.hpp"
+#include "expression/abstract_predicate_expression.hpp"
+#include "expression/expression_utils.hpp"
+#include "expression/logical_expression.hpp"
+#include "logical_query_plan/abstract_lqp_node.hpp"
+#include "statistics/cardinality_estimator.hpp"
+#include "utils/assert.hpp"
+
+namespace opossum {
+
+void JoinPredicateOrderingRule::apply_to(const std::shared_ptr<AbstractLQPNode>& node) const {
+  // Check if this is a multi predicate join.
+  if (node->type != LQPNodeType::Join || node->node_expressions.size() <= 1) {
+    _apply_to_inputs(node);
+    return;
+  }
+
+  const auto join_mode = std::static_pointer_cast<JoinNode>(node)->join_mode;
+
+  DebugAssert(cost_estimator, "JoinOrderingRule requires cost estimator to be set");
+  const auto caching_cardinality_estimator = cost_estimator->cardinality_estimator->new_instance();
+
+  // Estimate join selectivity of a predicate by getting cardinalities for a join node joining only on that one predicate.
+  //  “The join selectivity is the ratio “number of tuples in the result/number of tuples in the cartesian product.”
+  //  See http://www.vldb.org/journal/VLDBJ6/70060191.pdf for more infos on join selectivity.
+  auto predicate_cardinalities = std::unordered_map<std::shared_ptr<AbstractExpression>, Cardinality>{};
+  for (const auto& predicate : node->node_expressions) {
+    const auto single_predicate_join = JoinNode::make(join_mode, predicate);
+    single_predicate_join->set_left_input(node->left_input());
+    single_predicate_join->set_right_input(node->right_input());
+
+    predicate_cardinalities[predicate] = caching_cardinality_estimator->estimate_cardinality(single_predicate_join);
```

Comment:
precicate_cardinalities.emplace(predicate, caching_card....);

Reply:
As I understand it this should be exactly the same. Both insert a value constructed in-place into the map. From the documentation for std::map::operator[]:
"Inserts a value_type object constructed in-place from std::piecewise_construct, std::forward_as_tuple(key), std::tuple<>() if the key does not exist. This function is equivalent to return this->try_emplace(key).first->second;. (since C++17) "
Am I missing something?

Reply:
current method would call default constructor + assignment opertator, emplace would just call a single constructor, so should be preferred, afaict